src/bananas.js

```diff
@@ -0,0 +1,185 @@
+                if (node.initial_context === EXPR) {
+                    queue.push(node.filter.expr);
+                    predicates = node.filter.predicates;
+                    for (i = 0; i < predicates.length; i++) {
+                        queue.push(predicates[i]);
+                    }
+                }
+            } else if (node instanceof models.HashtagExpr) {
+                (function() {
+                    var oldToHashtag = node.toHashtag;
+                    node.toHashtag = function() {
+                        return transformFn(oldToHashtag());
+                    };
+                })();
+            }
+            children = node.getChildren();
+            for (i = 0; i < children.length; i++) {
+                queue.push(children[i]);
+                if (children[i].predicates && children[i].predicates.length) {
+                    predicates = children[i].predicates;
+                    for (j = 0; j < predicates.length; j++) {
+                        queue.push(predicates[j]);
+                    }
+                }
+            }
+        }
+        return parsedHashtags.toHashtag();
+    }
+
+    /*
+     * transforms bananas based on transformFn
+     */
+    function transform(input, transformFn) {
+        if (!input) { return input; }
+        input = getSymbols(input);
+        transformFn = transformFn || function (input) { return input; };
+        var state = OUTSIDE_BANANA,
+            strLen = input.length,
+            text = "",
+            currentReference = "";
+
+        for (var i = 0; i < strLen; i++) {
+            var current = input[i],
+                next = input[i+1];
+
+            if (state === OUTSIDE_BANANA) {
+                if (current === DELIMITER && next === DELIMITER) {
+                    text += DELIMITER;
+                    i++;
+                } else if (current === DELIMITER) {
+                    state = INSIDE_BANANA;
+                } else {
+                    text += current;
+                }
+            } else if (state === INSIDE_BANANA) {
+                if (current === DELIMITER && next === DELIMITER) {
+                    currentReference += DELIMITER;
+                    i++;
+                } else if (current === DELIMITER) {
+                    state = OUTSIDE_BANANA;
+                    text += transformFn(currentReference);
+                    currentReference = "";
+                } else if (next !== undefined){
+                    currentReference += current;
+                } else {
+                    // end of string, shouldn't happen, but will not
+                    // overestimate users or Vellum devs
+                    text += DELIMITER + currentReference;
+                    currentReference = "";
+                }
+            }
+        }
+
+        if (state === INSIDE_BANANA) {
+            // end of string, without the end delimiter
+            text += DELIMITER + currentReference;
```

Comment:
This is an error condition, correct? I think it should throw an exception (fail the transform) rather than guessing at the intent and trying to fix up broken input. It should fail in a way that developers can see that something is broken. Otherwise we have to support broken input forever and/or we'll be masking bugs in vellum. The same applies above in the `else { // end of string...` clause.